libgdf/src/gdf_table.cuh

```diff
@@ -45,6 +45,30 @@ struct ValidRange {
     }
 };
 
+template<typename size_type, typename byte_type>
+__device__
+byte_type get_bit(size_type index, byte_type const * data)
+{
+  constexpr uint32_t BITS_PER_BYTE = (8 * sizeof(byte_type));
+
+  size_type byte_index = index / BITS_PER_BYTE;
+  byte_type bit_offset = index % BITS_PER_BYTE;
+  byte_type mask = static_cast<byte_type>(1) << bit_offset;
+  return (data[byte_index] & mask) >> bit_offset;
+}
+
+template<typename size_type, typename byte_type>
+__device__
+void set_bit(size_type index, byte_type value, byte_type * data)
+{
+  constexpr uint32_t BITS_PER_BYTE = (8 * sizeof(byte_type));
+
+  size_type byte_index = index / BITS_PER_BYTE;
+  byte_type bit_offset = index % BITS_PER_BYTE;
+
+  // value can only be 0 or 1 i.e. 0x00 or 0x01
+  data[byte_index] |= (value << bit_offset);
```

Comment:
The way you are using this function, I believe it is safe that this operation is not done atomically. However, in general, this operation would need to be done with an `atomicOr` to avoid race-conditions from multiple threads trying to update the same mask concurrently. This function will be obsolete once a common set of bit utility functions are implemented (coming soon). I would add a TODO that this should be replaced once that is done.

Reply:
Furthermore, if `value` should only be `1` or `0`, then it's type should be `bool` and not `byte_type`.

Reply:
One last thing... `data`'s type should be `gdf_valid_type*`. And `BITS_PER_BYTE` should be replaced with `GDF_VALID_BITSIZE`: https://github.com/rapidsai/cudf/blob/master/libgdf/include/gdf/gdf.h#L10

Reply:
Since atomicOr would need a 32 bit value, I'd have to cast `data` to `uint32_t`. But allocations of gdf_valid_type is done in multiple of bytes, not multiple of 4 bytes. There's a danger that while setting the last values of data, I'll access illegal memory.

Reply:
For now, I'll remove `set_bit` method altogether to avoid it being used by someone else.

Reply:
Indeed, that's the bug you found in my other code :)

Luckily, `gdf_valid_type` will be updated soon to be a 4 or 8 byte type, and we won't have these problems.